sycl/include/sycl/esimd/detail/esimd_intrin.hpp

```diff
@@ -0,0 +1,290 @@
+//==------------ esimd_intrin.hpp - DPC++ Explicit SIMD API   --------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Declares Explicit SIMD intrinsics used to implement working with
+// the SIMD classes objects.
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <cstdint>
+#include <sycl/esimd/detail/esimd_types.hpp>
+#include <sycl/esimd/esimd_enum.hpp>
+
+// \brief __esimd_rdregion: region access intrinsic.
+//
+// @param Ty the element data type, one of i8, i16, i32, i64, half, float,
+// double. In particular bool (i1) and pointer types are not allowed.
+//
+// @param N the input vector size.
+//
+// @param M the return vector size.
+//
+// @param VStride the vertical stride in elements between rows.
+//
+// @param Width the size or each row, non-zero and even divides `M`.
+//
+// @param Stride horizontal stride in elements within each row.
+//
+// @param ParentWidth the width of the input vector when viewed as a 2D
+// matrix. Ignored if offset is a constant.
+//
+// @param Input the input vector
+//
+// @param Offset the starting offset in bytes.
+//
+// @return the region extracted.
+//
+// This intrinsic computes a vector Result:
+//
+// \code{.cpp}
+// uint16_t EltOffset = Offset / sizeof(Ty);
+// assert(Offset % sizeof(Ty) == 0);
+//
+// int NumRows = M / Width;
+// assert(M % Width == 0);
+//
+// int Index = 0;
+// for (int i = 0; i < NumRows; ++i) {
+//   for (int j = 0; j < Width; ++j) {
+//     Result[Index++] = Input[i * VStride +  j * Stride +
+//     EltOffset];
+//   }
+// }
+// \endcode
+//
+template <typename Ty, int N, int M, int VStride, int Width, int Stride,
```

Comment:
Nit: Is there any reason to use `Ty` here instead of the more standard `T`?

Reply:
It's just naming convention difference, can be changed to align with open source preference.

Reply:
Renamed in the whole file.

Reply:
Fixed.

Reply:
@DenisBakhvalov Did you undo these changes in a later commit?  I'm still seeing lots of `Ty` templates, and I'm wondering if this was intentional.

Reply:
No, I did not. :) But I changed template names in `esimd_intrin.hpp` only. We can add an incremental patch to fix other places later.